artiq/frontend/aqctl_proxy_moninj.py

```diff
@@ -0,0 +1,298 @@
+#!/usr/bin/env python3
+
+import argparse
+import asyncio
+import atexit
+import logging
+from abc import ABC, abstractmethod
+
+from sipyco import common_args
+from sipyco.asyncio_tools import atexit_register_coroutine
+from sipyco.pc_rpc import Server as RPCServer
+from sipyco.sync_struct import Publisher, Notifier, Subscriber
+
+from artiq import __version__ as artiq_version
+from artiq.coredevice.comm_moninj import CommMonInj
+
+logging.basicConfig(
+    level=logging.DEBUG,
+    format="%(asctime)s [%(levelname)s] %(message)s",
+    handlers=[
+        logging.StreamHandler()
+    ]
+)
+logger = logging.getLogger(__name__)
+
+
+class Connectable(ABC):
+    def __init__(self):
+        self.connected_event = asyncio.Event()
+
+    @abstractmethod
+    async def connect(self, addr, port): pass
+
+    @abstractmethod
+    async def disconnect(self): pass
+
+    async def on_connected(self):
+        self.connected_event.set()
+
+    async def on_disconnected(self):
+        self.connected_event.clear()
+
+    @property
+    def connected(self):
+        return self.connected_event.is_set()
+
+
+class HasSupervisor(ABC):
+    @property
+    @abstractmethod
+    def supervisor(self): pass
+
+
+class HasCoreAddr(ABC):
+    @property
+    @abstractmethod
+    def core_addr(self): pass
```

Comment:
Are those little classes really useful?

Reply:
It is useful that the event handlers can access the properties of "child class". My intention is to make it look like a Rust trait (also to emulate composition over inheritance), but it seems like there is no way to do that in Python elegantly.

Reply:
Please just keep it simple - this is supposed to be just a little program, no need to invent new Python programming techniques here.

Reply:
Still, this cleanly separate the callback mixin class while still having access to parent class. I'd suggest keeping it.